examples/vision/fixres.py

```diff
@@ -0,0 +1,319 @@
+results.
+"""
+
+# Reference: https://github.com/facebookresearch/FixRes/blob/main/transforms_v2.py.
+
+batch_size = 128
+auto = tf.data.AUTOTUNE
+smaller_size = 128
+bigger_size = 224
+
+size_for_resizing = int((bigger_size / smaller_size) * bigger_size)
+central_crop_layer = layers.CenterCrop(bigger_size, bigger_size)
+
+
+def make_dataset(
+    dataset: tf.data.Dataset,
+    train: bool,
+    image_size: int = smaller_size,
+    fixres: bool = True,
+):
+    def preprocess_initial(image, label):
+        """
+        Initial preprocessing function for training on smaller resolution.
+        For training, do random_horizontal_flip -> random_crop.
+        For validation, just resize.
+        No color-jittering has been used.
+        """
+        if train:
+            channels = image.shape[-1]
+            begin, size, _ = tf.image.sample_distorted_bounding_box(
+                tf.shape(image),
+                tf.zeros([0, 0, 4], tf.float32),
+                area_range=(0.05, 1.0),
+                min_object_covered=0,
+                use_image_if_no_bounding_boxes=True,
+            )
+            image = tf.slice(image, begin, size)
+
+            image.set_shape([None, None, channels])
+            image = tf.image.resize(image, [image_size, image_size])
+            image = tf.image.random_flip_left_right(image)
+        else:
+            image = tf.image.resize(image, [image_size, image_size])
+
+        image = (image - 127.5) / 127.5
+        return image, label
+
+    def preprocess_finetune(image, label):
+        """
+        Preprocessing function for fine-tuning on a higher resolution.
+        For training, resize to a bigger resolution to maintain the ratio ->
+            random_horizontal_flip -> center_crop.
+        For validation, do the same without any horizontal flipping.
+        No color-jittering has been used.
+        """
+        image = tf.image.resize(image, [size_for_resizing, size_for_resizing])
+        if train:
+            image = tf.image.random_flip_left_right(image)
+        image = central_crop_layer(image[None, ...])[0]
+        image = (image - 127.5) / 127.5
+
+        return image, label
+
+    if image_size not in [smaller_size, bigger_size]:
+        raise ValueError(f"{image_size} resolution is not supported.")
+
+    # Determine which preprocessing function we are using.
+    if image_size == smaller_size:
+        preprocess_func = preprocess_initial
+    elif not fixres and image_size == bigger_size:
+        preprocess_func = preprocess_initial
+    else:
+        preprocess_func = preprocess_finetune
+
+    if train:
+        dataset = dataset.shuffle(batch_size * 10)
+
+    return dataset.map(preprocess_func, auto).batch(batch_size).prefetch(auto)
+
+
+"""
+Notice how the augmentation transforms vary for the kind of dataset we are preparing.
+"""
+
+"""
+## Prepare datasets
+"""
+
+initial_train_dataset = make_dataset(train_dataset, train=True, image_size=smaller_size)
```

Comment:
Perhaps the class approach would make most sense since you wouldn't have to pass the `image_size` arg to 6 different function calls (it would be a constructor arg)